File: backend/src/main/java/com/urmyfood/backend/domain/repository/VoucherRepository.java
package com.urmyfood.backend.domain.repository;

import com.urmyfood.backend.domain.model.Voucher;
import java.util.List;
import java.util.Optional;

public interface VoucherRepository {
    List<Voucher> findAllActive();
    List<Voucher> findSavedByCustomerId(Long customerId);
    Optional<Voucher> findById(Long id);
    Optional<Voucher> findByCode(String code);
    boolean isSaved(Long voucherId, Long customerId);
    void saveVoucher(Long voucherId, Long customerId);
    void unsaveVoucher(Long voucherId, Long customerId);
    Voucher save(Voucher voucher);
    void deleteById(Long id);
    List<Voucher> findAll();
}

File: frontend/app-user/src/main/java/com/urmyfood/user/domain/repository/VoucherRepository.kt
package com.urmyfood.user.domain.repository

import com.urmyfood.user.data.model.VoucherResponse
import com.urmyfood.user.domain.model.Result

interface VoucherRepository {
    suspend fun getActiveVouchers(token: String?): Result<List<VoucherResponse>>
    suspend fun getSavedVouchers(token: String): Result<List<VoucherResponse>>
    suspend fun saveVoucher(token: String, voucherId: Long): Result<com.urmyfood.user.data.model.SavedVoucherResponse>
    suspend fun unsaveVoucher(token: String, voucherId: Long): Result<com.urmyfood.user.data.model.SavedVoucherResponse>
}
